tensorboard/plugins/graph/tf_graph_common/common_test.ts

```diff
@@ -0,0 +1,124 @@
+      'setMessage',
+      'updateProgress',
+      'reportError',
+    ]);
+  });
+
+  it('notifies listeners of events when graph loads', async () => {
+    // Use mock data instead of fetching from a file. This may skip parser
+    // related events.
+    spyOn(tf_graph_parser, 'fetchAndParseGraphData').and.returnValue(
+      Promise.resolve({
+        node: [{name: 'add1', input: [], device: '', op: 'add', attr: []}],
+        versions: [],
+        library: {function: []},
+      })
+    );
+
+    const debugListenerSpy = spyOn(tb_debug, 'notifyActionEventFromPolymer');
+    await tf_graph_loader.fetchAndConstructHierarchicalGraph(
+      mockTracker,
+      null /* remotePath */,
+      null /* pbTxtFile */
+    );
+
+    const firstArgs = debugListenerSpy.calls
+      .allArgs()
+      .map((args: any[]) => args[0]);
+    expect(firstArgs).toEqual([
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.NORMALIZING_NAMES,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.BUILD_SLIM_GRAPH,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.HIERARCHY_ADD_NODES,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.HIERARCHY_DETECT_SERIES,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.HIERARCHY_ADD_EDGES,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction:
+          tb_debug.GraphDebugEventId.HIERARCHY_FIND_SIMILAR_SUBGRAPHS,
+        eventValue: jasmine.any(Number),
+      },
+      {
+        eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
+        eventAction: tb_debug.GraphDebugEventId.GRAPH_LOAD_SUCCEEDED,
+        eventValue: jasmine.any(Number),
+      },
+    ]);
+  });
+
+  it('notifies listeners of events when graph fails to load', async () => {
+    // Use mock data instead of fetching from a file. This may skip parser
+    // related events.
+    spyOn(tf_graph_parser, 'fetchAndParseGraphData').and.returnValue(
+      Promise.resolve({
+        // Graphs with no 'node' objects are invalid.
+        node: undefined,
+        versions: [],
+        library: {function: []},
+      })
+    );
+
+    const debugListenerSpy = spyOn(tb_debug, 'notifyActionEventFromPolymer');
+
+    try {
+      await tf_graph_loader.fetchAndConstructHierarchicalGraph(
+        mockTracker,
+        null /* remotePath */,
+        null /* pbTxtFile */
+      );
```

Comment:
I think you should fail the test if it does not throw?
```ts
expect(async () => await tf_graph_loader.fetchAndConstructHierarchicalGraph(
        mockTracker,
        null /* remotePath */,
        null /* pbTxtFile */
      )).toThrowError();
const firstArgs = debugListenerSpy.calls
        .allArgs()
        .map((args: any[]) => args[0]);
      expect(firstArgs).toEqual([
        {
          eventCategory: tb_debug.GRAPH_DEBUG_TIMING_EVENT_CATEGORY,
          eventAction: tb_debug.GraphDebugEventId.GRAPH_LOAD_FAILED,
          eventValue: jasmine.any(Number),
        },
      ]);
```

Above should work, I think.

Reply:
Good catch.  The example given unfortunately did not work because the 2nd `expect` call synchronously fires after the first `expect`, but before the fetchAndConstructHierarchicalGraph finishes.

Done using a `await expectAsync(promise).toBeRejected()` pattern.